README.md

```diff
@@ -1,526 +1,917 @@
+    return value;
+  }
 }
 
-const e = new Element();
-e.increment();  // logs 1
-e.increment();  // logs 2
+class Child {}
+class OtherChild {}
+
+class Parent {
+  @register child1 = new Child();
+  @register child2 = new OtherChild();
+}
+
+let parent = new Parent();
+getChildren(parent); // [Child, OtherChild]
 ```
 
-When fields are decorated, the "wrapped" value is an object with two properties: `get` and `set` functions that manipulate the underlying storage. They are built to be `.call()`ed with the instance of the class as a receiver. The decorator can then return a new object of the same form. (If one of the callbacks is missing, then it is left in place unwrapped.)
+#### Classes
 
-```mjs
-// tracked.mjs
+```ts
+type ClassDecorator = (value: Function, context: {
+  kind: "class";
+  name: string;
+  defineMetadata(key: string | symbol | number, value: unknown);
+}) => Function | void;
+```
 
-export function tracked({get, set}) {
-  return {
-    get,
-    set(value) {
-      if (get.call(this) !== value) {
-        set.call(this, value);
-        this.render();
+Class decorator's receive the class that is being decorated as the first parameter, and may optionally return a new class to replace it. If a non-constructable value is returned, then an error is thrown.
+
+We can further extend our `@logged` decorator to log whenever an instance of a class is created:
+
+```js
+function logged(value, { kind, name }) {
+  if (kind === "class") {
+    return class extends value {
+      constructor(...args) {
+        console.log(`constructing an instance of ${name} with arguments ${args.join(", ")}`);
       }
     }
-  };
+  }
+
+  // ...
 }
+
+@logged
+class C {}
+
+new C(1);
+// constructing an instance of C with arguments 1
 ```
 
-This example could be roughly desugared as follows:
+This example roughly "desugars" to the following (i.e., could be transpiled as such):
 
-```mjs
-let initialize, get, set;
+```js
+class C {}
 
-class Element {
-  #counter = initialize.call(this, 0);
-  get counter() { return this.#counter; }
-  set counter(v) { this.#counter = v; }
+C = logged(C, {
+  kind: "class",
+  name: "C",
+  defineMetadata() { /**/ }
+}) ?? C;
 
-  increment() { this.counter++; }
+new C(1);
+```
 
-  render() { console.log(counter); }
-}
+### New Class Elements
 
-{ get, set } = Object.getOwnPropertyDescriptor(Element.prototype, "counter");
-{ get, set, initialize } = tracked({get, set}, { kind: "field", name: "counter", isStatic: false })
-Object.defineProperty(Element.prototype, "counter", {get, set});
+#### Class Properties
+
+Class properties are a new construct, defined by adding the `prop` keyword in front of a class field:
+
+```js
+class C {
+  prop x = 1;
+}
 ```
 
-### Limited access to private fields and methods
+Properties, unlike fields, define a getter and setter on the class prototype. This getter and setter default to getting and setting a value on a private slot. The above roughly desugars to:
```

Comment:
what is the point of doing this? How is an undecorated `prop` any different than a public field? (i suppose it does a brand check, but is there a use case for that?)

Reply:
Right, the thinking is that `prop` fundamentally changes the semantics of a field, and then decorators layer naturally on top of the new semantics. While the semantics on their own have limited use cases, there are many use cases for decorators with these semantics. Given the other constraints around shape changes and static analyzability, this option keeps the application of decorators consistent while solving all of the use cases.

Reply:
Separate from the name (which is imo a nonstarter), I don't think we should be adding fundamental field semantics in the absence of decorators for imagined use cases.

I think `@prop: decorator` and `@init: decorator` are indeed the same in this sense, and i'd prefer to see both of them remain exclusively attached to decorator syntax.

Reply:
The difficulty comes with combining `@prop: decorator` and `@decorator` declarations then. Does applying one `@prop: decorator` then turn the field into a prop-field, and all other decorators now need to handle that prop-field? Likewise with init-methods.

This to me signals that the `@prop:` style modifier is fundamentally in the wrong location. It is not actually modifying the decorator, it is modifying the element itself, so it is correct to place the modifier on the element, regardless of whether it has utility on its own.

Reply:
I'm not sure what you mean by "decorator declarations". But yes, if you compose decorators, and one of them adds `prop`, then the rest need to handle that, just like if a class decorator changes the `.name` or identity of a class, the rest of the decorators need to handle that.

They're all modifying the element themselves. The extra syntax is because the application of a single decorator has modified semantics.

Reply:
Another thing to note is that the current location would compose nicely with the proposed [Grouped Accessors and Auto-Accessors](https://github.com/rbuckton/proposal-grouped-and-auto-accessors) proposal. Like `prop`, users would be able to modify a field declaration with `{ get; set; }` to create getters and setters on it. `prop` would effectively become syntactic sugar for this style of declaration.

The reason we decided not to use the auto-accessor syntax was so that we aren't blocked on that proposal, since it is currently in stage 1, and because the syntactic sugar of the keyword was seen as a good thing among users who will have to use `prop` to convert their existing decorators.

Reply:
> But yes, if you compose decorators, and one of them adds prop, then the rest need to handle that, just like if a class decorator changes the .name or identity of a class, the rest of the decorators need to handle that.

So let's dig in a bit more there. If this is the case, then all three of these would be a valid way to get a prop, and effectively be equivalent, no?

```js
@prop: a @b @c x = 123;
@a @prop: b @c x = 123;
@a @b @prop: c x = 123;

// Can also call it on all of them
@prop: a @prop: b @prop: c x = 123
```

It seems like this creates more confusion than it helps, again because the transform is not coming from the decorator itself, its coming from the base element being changed.

Reply:
It's coming from the way the decorator is called, essentially, yes - just like the `this` value and the arguments of a function do.

Reply:
hmm, I think there may be a bit of a miscommunication. Are you suggesting that the value would only change _after_ the `@prop:` decorator? e.g.

```js
@a // prop
@prop: b //prop
@c // field
x = 123;
```

That feels like it would make a bit more sense. I still think it would be a major refactoring hazard though

Reply:
Another possibility would be to allow `prop` and `field` decorators to interleave, and call field decorators as field decorators instead:

```js
@a // kind: 'field'
@prop:b // kind: 'prop'
@c // kind: 'field'
x = 123;
```

I _think_ this might work, since the field decorators would just be returning initializer functions, which could be tacked onto the private slot for the prop